dsl/src/main/scala/org/http4s/dsl/impl/Path.scala

```diff
@@ -98,6 +98,40 @@ object -> {
     Some((req.method, req.pathInfo))
 }
 
+object ->> {
+  private val allMethods = Method.all.toSet
+
+  /** Extractor to match an http resource and then enumerate all supported methods:
+    * {{{
+    *   (request.method, Path(request.path)) match {
+    *     case withMethod ->> Root / "test.json" => withMethod {
+    *       case Method.GET => ...
+    *       case Method.POST => ...
+    * }}}
+    *
+    * Returns an error response if the method is not matched, in accordance with [[https://datatracker.ietf.org/doc/html/rfc7231#section-4.1 RFC7231]]
+    */
+  def unapply[F[_]: Applicative](
+      req: Request[F]): Some[(PartialFunction[Method, F[Response[F]]] => F[Response[F]], Path)] =
+    Some {
+      (
+        pf =>
+          pf.applyOrElse(
+            req.method,
+            (method: Method) =>
+              Applicative[F].pure {
+                if (allMethods.contains(method)) {
+                  Response(
+                    status = Status.MethodNotAllowed,
+                    headers = Headers(Allow(allMethods.filter(pf.isDefinedAt))))
+                } else { Response(status = Status.NotImplemented) }
+              }
+          ),
+        req.pathInfo)
+
+    }
```

Comment:
I like scalafmt, but if I didn't, this would be in my slide deck. 😆

Reply:
It was never gonna be pretty but I like the fact that scalafmt makes that not be my fault.

Reply:
By the way, I'm not super happy that the Allow header has to be regenerated on every match failure but I can't see a way around that as long as the DSL is matcher-based.